File: QuickCar/Core/Fuel/FuelStatistics.swift
import Foundation
import SwiftData

@MainActor
final class FuelStatistics {
    private let modelContext: ModelContext
    private let settings: AppSettings

    init(modelContext: ModelContext, settings: AppSettings) {
        self.modelContext = modelContext
        self.settings = settings
    }

    func summary(for range: DateInterval) -> DrivingSummary {
        let descriptor = FetchDescriptor<Trip>(
            predicate: #Predicate { trip in
                trip.startedAt >= range.start && trip.startedAt < range.end
            }
        )
        let trips = (try? modelContext.fetch(descriptor)) ?? []
        return DrivingSummary(trips: trips, pricePerLiter: settings.pricePerLiter)
    }

    func dailyFuel(lastDays: Int = 14) -> [(date: Date, liters: Double)] {
        let calendar = Calendar.current
        let end = calendar.startOfDay(for: Date()).addingTimeInterval(86400)
        guard let start = calendar.date(byAdding: .day, value: -lastDays, to: end) else { return [] }
        let descriptor = FetchDescriptor<Trip>(
            predicate: #Predicate { trip in
                trip.startedAt >= start && trip.startedAt < end
            }
        )
        let trips = (try? modelContext.fetch(descriptor)) ?? []
        var buckets: [Date: Double] = [:]
        for trip in trips {
            let day = calendar.startOfDay(for: trip.startedAt)
            buckets[day, default: 0] += trip.fuelUsedL
        }
        return (0..<lastDays).compactMap { offset in
            guard let day = calendar.date(byAdding: .day, value: -(lastDays - 1 - offset), to: calendar.startOfDay(for: Date())) else { return nil }
            return (day, buckets[day] ?? 0)
        }
    }

    func lastRefuel() -> RefuelEntry? {
        var descriptor = FetchDescriptor<RefuelEntry>(sortBy: [SortDescriptor(\.date, order: .reverse)])
        descriptor.fetchLimit = 1
        return try? modelContext.fetch(descriptor).first
    }

    func measuredVsEstimated() -> (measured: Double?, estimated: Double?) {
        var descriptor = FetchDescriptor<RefuelEntry>(
            predicate: #Predicate { $0.isFullTank },
            sortBy: [SortDescriptor(\.date, order: .reverse)]
        )
        descriptor.fetchLimit = 2
        let fulls = (try? modelContext.fetch(descriptor)) ?? []
        guard fulls.count >= 2,
              let odoNew = fulls[0].odometerKm,
              let odoOld = fulls[1].odometerKm,
              odoNew > odoOld else {
            return (nil, nil)
        }
        let delta = odoNew - odoOld
        let measured = fulls[0].liters / delta * 100
        let range = DateInterval(start: fulls[1].date, end: fulls[0].date)
        let estimated = summary(for: range).avgL100
        return (measured, estimated > 0 ? estimated : nil)
    }
}

File: mobile/src/core/fuel/fuelStatistics.ts
import { tripRepository, type DateInterval } from "../storage/tripRepository";
import { fuelRepository } from "../storage/fuelRepository";
import { emptyDrivingSummary, summarize, type DrivingSummary, type RefuelEntry } from "../storage/models";

export class FuelStatistics {
  async summary(range: DateInterval, pricePerLiter: number): Promise<DrivingSummary> {
    const trips = await tripRepository.trips(range);
    return trips.length ? summarize(trips, pricePerLiter) : emptyDrivingSummary();
  }

  async dailyFuel(lastDays = 14): Promise<Array<{ date: number; liters: number }>> {
    const end = new Date();
    end.setHours(0, 0, 0, 0);
    end.setDate(end.getDate() + 1);
    const start = new Date(end);
    start.setDate(start.getDate() - lastDays);
    const trips = await tripRepository.trips({ start: start.getTime(), end: end.getTime() });
    const buckets = new Map<number, number>();
    for (const trip of trips) {
      const d = new Date(trip.startedAt);
      d.setHours(0, 0, 0, 0);
      const key = d.getTime();
      buckets.set(key, (buckets.get(key) ?? 0) + trip.fuelUsedL);
    }
    const result: Array<{ date: number; liters: number }> = [];
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    for (let offset = 0; offset < lastDays; offset++) {
      const day = new Date(today);
      day.setDate(day.getDate() - (lastDays - 1 - offset));
      result.push({ date: day.getTime(), liters: buckets.get(day.getTime()) ?? 0 });
    }
    return result;
  }

  async lastRefuel(): Promise<RefuelEntry | undefined> {
    const all = await fuelRepository.allRefuels();
    return all[0];
  }

  async measuredVsEstimated(): Promise<{ measured?: number; estimated?: number }> {
    const pair = await fuelRepository.lastFullTankPair();
    if (!pair) return {};
    const [latest, previous] = pair;
    if (latest.odometerKm == null || previous.odometerKm == null || !(latest.odometerKm > previous.odometerKm)) {
      return {};
    }
    const delta = latest.odometerKm - previous.odometerKm;
    const measured = (latest.liters / delta) * 100;
    const summary = await this.summary({ start: previous.date, end: latest.date }, 0);
    return { measured, estimated: summary.avgL100 > 0 ? summary.avgL100 : undefined };
  }
}

export const fuelStatistics = new FuelStatistics();
